oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java

```diff
@@ -96,7 +96,7 @@ public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2Authentic
 		OAuth2AccessToken token = userRequest.getAccessToken();
 		Map<String, Object> attributes = this.attributesConverter.convert(userRequest).convert(response.getBody());
 		Collection<GrantedAuthority> authorities = getAuthorities(token, attributes, userNameAttributeName);
-		return new DefaultOAuth2User(authorities, attributes, userNameAttributeName);
+		return new DefaultOAuth2User(attributes.get(userNameAttributeName).toString(), attributes, authorities);
```

Comment:
Is it guaranteed at this point that attribute.get(userNameAttributeName) is non-null? If not, please do a null check so that the code can offer a more informative error message.

Reply:
You're right, `attributes` is nullable and an assertion is appropriate here, basically what my newly introduced `DefaultOAuth2User.getNameFromAttributes` is doing.

However, I can't help but feel that "get name from attributes using nameAttributeKey with assertions" is a frequent enough operation (4 mostly similar occurrences have been touched by this PR) to maybe warrant a separate method.

What would you think about changing the constructors to use `name` instead of `nameAttributeKey` (as done by the PR) but introducing static factory methods for `DefaultOidcUser` and `DefaultOAuth2User` taking `nameAttributeKey` as parameter instead of `name`?  
That would get rid of the old, inconsistent constructors (in terms of both the parameter order and the usage of `nameAttributeKey` instead of `name`) while giving a clear upgrade path. Additionally, `OidcUserAuthority.collectClaims` mentioned above would not need to be duplicated or made accessible.

Reply:
I've added builders for `DefaultOAuth2User` and `DefaultOidcUser`. Right now the new "canonical constructor" is still public, but it might be better to make it private to be in line with, e.g., `OAuth2AuthorizeRequest`?